Flaky tests

o- opened the pull request.

o- committed:
move flaky tests to one target

o- committed:
further increase max compilation size

o- committed:
auto retry flaky tests

o- committed:
update retry

o- committed:
fix promise flowing into asTest

The type we attribute to the cast in the inliner was totally wrong.
Because the wrapper had an eager type, we would let the promise
flow into non-promise instructions, instead of adding a force
instruction and reading out the eager value.

o- commented on rir/src/compiler/opt/inline.cpp:
```diff
@@ -200,14 +200,11 @@ class TheInliner {
-                                    mk->isEager()
-                                        ? mk->eagerArg()->type.forced()
-                                        : ld->type);
+                                auto cast =
```
@mhyee that was the bug you where seeing.

when we inline we do the following:
```
   prom  %1 = mkarg(...)
   val^   %2 = Cast(val^, %1)
```

we add a cast to convert a *promise-as-a-value* to a *lazy-value*. The former is a promise in value position, ie. we mean the promise itself. The later is a lazy value, ie. we mean the value we get out of the promise. That cast is needed, because normally the call instruction does this implicit "conversion".

Now I tried to be clever and already propagate the underlying type, if the argument was eager. But that caused the compiler to think that the result of `Cast` is a value without promise wrapper, when in fact it has a promise wrapper, that needs to be removed.

Most of the time this does not cause any issues, since many operations just implicitly evaluate promises. But the `asbool_` instruction does not do so...

o- committed:
CastType: propagate the underlying type, but the correct one

the initial idea was good, but we need to mark the type such that
we remember that there is still a promise wrapped there.

mhyee commented:
I stumbled upon this because I changed the following:

https://github.com/reactorlabs/rir/blob/7c13298568cf587f933e19fcade1ac8be70dbd85/rir/src/compiler/pir/instruction.h#L902L905
```cpp
    Is(uint32_t sexpTag, Value* v)
        : FixedLenInstruction(PirType(RType::logical).scalar(),
                              {{PirType::val()}}, {{v}}), sexpTag(sexpTag) {}
```

Should the type be `NativeType::test` instead of `PirType(RType::logical).scalar()`? When I made that change, that's when the `asbool_` thing showed up.

o- committed:
update submodule: fix some test failing because they take longer

o- commented:
> Should the type be NativeType::test instead of PirType(RType::logical).scalar()?

if you look at the implementation of `is_` it only returns true/false, so test is more precise. but of course then the problem is that if you want to use the instruction in a situation where an rtype is requested, you have a problem... it seems to me that this instruction could be either or. certainly `PirType(RType::logical).scalar()` does not feel wrong (and I suspect that it would not break on top of this branch anymore). otoh, you can't directly use it as branch input anymore (it will add a useless asbool).

in some cases we use the same instruction but just change the type. the type of an instruction is mutable and this seems like one of the instructions where i'd say, make the type either or, dending on where you want to feed it to.

o- committed:
remove redundancy in flaky test target

o- committed:
fix for an impossible to debug problem

the problem was mk_env_ reordering, which caused the wrong environment
to be created first and thus resulted in a context with a wrong env.

For example in:

```
e1 = mkenv
     pushcontext
e2 = mkenv
```

we would reorder to:

```
     pushcontext
e1 = mkenv
e2 = mkenv
```

which causes the env e1 to be stored on the inner context and e2
nowhere.

To fix the issue I added a `context` number to mkenv. Thus we get:

```
e1 = mkenv         context 1
     pushcontext
e2 = mkenv         context 1
```

and after reordering

```
     pushcontext
e1 = mkenv         context 2
e2 = mkenv         context 1
```

as can be seen we update the context position, when hoisting over a
pushcontext (or when removing push/pop contexts). Thus the mkenv always
updates the correct environment.

This also solves the issue where a deoptimization recreates an
envrionemnt. For example we sometimes have:

```
  cp1 = checkpoint  DEOPT | CONT

DEOPT:
  e1' = mkenv     context 1
  deopt

CONT:
  e1 = mkenv      context 1
  assume(cp1, ...)
```

here we create the env `e1` twice in the deopt case. Since we are now
precise in tracking the contexts a mkenv belongs to, we can safely
override `e1` on the context with `e1'` in the deopt case.
